src/core/reference/include/openvino/reference/nms_rotated_util.hpp

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) Facebook, Inc. and its affiliates.
+// The implementation for rotated boxes intersection is based on the code from:
+// https://github.com/facebookresearch/detectron2/blob/v0.6/detectron2/layers/csrc/box_iou_rotated/box_iou_rotated_utils.h
+
+#pragma once
+
+#include <assert.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <iostream>
+#include <iterator>
+#include <numeric>
+#include <vector>
+
+namespace iou_rotated {
+struct RotatedBox {
+    float x_ctr, y_ctr, w, h, a;
+};
+struct Point {
+    float x, y;
+    Point(const float& px = 0, const float& py = 0) : x(px), y(py) {}
+    Point operator+(const Point& p) const {
+        return Point(x + p.x, y + p.y);
+    }
+    Point& operator+=(const Point& p) {
+        x += p.x;
+        y += p.y;
+        return *this;
+    }
+    Point operator-(const Point& p) const {
+        return Point(x - p.x, y - p.y);
+    }
+    Point operator*(const float coeff) const {
+        return Point(x * coeff, y * coeff);
+    }
+};
+
+static inline float dot_2d(const Point& A, const Point& B) {
+    return A.x * B.x + A.y * B.y;
+}
+
+static inline float cross_2d(const Point& A, const Point& B) {
+    return A.x * B.y - B.x * A.y;
+}
+
+static inline void get_rotated_vertices(const RotatedBox& box, Point (&pts)[4]) {
+    // M_PI / 180. == 0.01745329251
+    double theta = box.a;  // angle already in radians
+    float cosTheta2 = (float)cos(theta) * 0.5f;
+    float sinTheta2 = (float)sin(theta) * 0.5f;
+
+    // y: top --> down; x: left --> right
+    pts[0].x = box.x_ctr - sinTheta2 * box.h - cosTheta2 * box.w;
+    pts[0].y = box.y_ctr + cosTheta2 * box.h - sinTheta2 * box.w;
+    pts[1].x = box.x_ctr + sinTheta2 * box.h - cosTheta2 * box.w;
+    pts[1].y = box.y_ctr - cosTheta2 * box.h - sinTheta2 * box.w;
+    pts[2].x = 2 * box.x_ctr - pts[0].x;
+    pts[2].y = 2 * box.y_ctr - pts[0].y;
+    pts[3].x = 2 * box.x_ctr - pts[1].x;
+    pts[3].y = 2 * box.y_ctr - pts[1].y;
+}
+
+static inline int get_intersection_points(const Point (&pts1)[4], const Point (&pts2)[4], Point (&intersections)[24]) {
```

Comment:
General comment function make some comment what they are doing. It could give better overview what is intention for implementation.
I'm not sure what exactly means get intersection points? But there can be intersection point of two lines.

Reply:
Added more comments, the intersection points are the points defining boxes intersection area (intersection points of the edges and the corners within the boxes area). General algorithm can be also found in the spec [NMSRotated_13.md](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sort/NMSRotated_13.md).